{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8062d3fa-d2d8-459e-b537-a366df4b4cda",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from snakemake.io import expand\n",
    "import yaml\n",
    "import numpy as np\n",
    "from utils import *\n",
    "from sm_utils import *\n",
    "from bc_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "00b8270f-5197-452e-a741-a549d149b3af",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = '../configs/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55943c30-8307-4268-a899-c36277d9d060",
   "metadata": {},
   "source": [
    "## 231014 add klue genotype info to our obs table\n",
    "## use a heuristic to call the genotype of each multiplexed cell\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "2720a846-8675-4964-a739-531999249985",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = '../kallisto/igvf_010/Sublibrary_2/filtered/adata.h5ad'\n",
    "genotypes = '../klue/igvf_010/Sublibrary_2/genotype_counts.tsv'\n",
    "ofile = '../klue/igvf_010/Sublibrary_2/adata.h5ad'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "dd3d82ea-9849-40bd-a4fc-7ca472af3eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genotypes():\n",
    "    g = ['WSBJ','NZOJ',\n",
    "         'B6J','NODJ','129S1J',\n",
    "         'CASTJ','AJ','PWKJ']\n",
    "    return g\n",
    "    \n",
    "def assign_demux_genotype(df):\n",
    "    \"\"\"\n",
    "    Assigns a cell the genotype w/ the maximum of counts\n",
    "    between the two genotypes that were loaded in the well\n",
    "    the cell is from.\n",
    "    \n",
    "    Parameters:\n",
    "        df (pandas DataFrame): DF of obs table for each cell w/ \n",
    "            klue counts for each genotype and multiplexed genotype\n",
    "            columns\n",
    "    \"\"\"\n",
    "    genotype_cols = get_genotypes()\n",
    "    \n",
    "    # restrict to nuclei w/ genetic multiplexing\n",
    "    df = df.loc[df.well_type=='Multiplexed'].copy(deep=True)\n",
    "    \n",
    "    # fill nans once again\n",
    "    df[genotype_cols] = df[genotype_cols].fillna(0)\n",
    "    \n",
    "    # loop through each multiplexed genotype combo\n",
    "    # use those genotypes to determine which, \n",
    "    # between the two, has the highest counts\n",
    "    keep_cols = ['mult_genotype',\n",
    "                 'mult_genotype_1',\n",
    "                 'mult_genotype_2']+genotype_cols\n",
    "    df = df[keep_cols]\n",
    "    temp2 = pd.DataFrame()\n",
    "    for g in df.mult_genotype.unique().tolist():\n",
    "        # print(g)\n",
    "        temp = df.loc[df.mult_genotype==g].copy(deep=True)\n",
    "\n",
    "        g1 = temp.mult_genotype_1.unique().tolist()\n",
    "        assert len(g1) == 1\n",
    "        g1 = g1[0]\n",
    "\n",
    "        g2 = temp.mult_genotype_2.unique().tolist()\n",
    "        assert len(g2) == 1\n",
    "        g2 = g2[0]\n",
    "    \n",
    "        # find the best match and report ties if the \n",
    "        # values are the same\n",
    "        temp['new_genotype'] = temp[[g1,g2]].idxmax(axis=1)\n",
    "        temp.loc[temp[g1]==temp[g2], 'new_genotype'] = 'tie'\n",
    "\n",
    "        temp2 = pd.concat([temp2, temp], axis=0)\n",
    "\n",
    "    df = df.merge(temp2['new_genotype'], how='left',\n",
    "                  left_index=True, right_index=True)\n",
    "    df = df[['new_genotype']]\n",
    "    \n",
    "    assert len(df.loc[df.new_genotype.isnull()].index) == 0\n",
    "\n",
    "    return df\n",
    "\n",
    "def merge_kallisto_klue(f, genotypes, ofile):\n",
    "    \"\"\"\n",
    "    Merge in the klue results with the kallisto results. Use\n",
    "    a heuristic (which should be changeable / is subject to change)\n",
    "    to determine the genotype assignment to each cell\n",
    "    \"\"\"\n",
    "    \n",
    "    adata = sc.read(f)\n",
    "    df = pd.read_csv(genotypes, sep='\\t')\n",
    "    df.set_index('cellID', inplace=True)\n",
    "    \n",
    "    # make sure we won't dupe any cols\n",
    "    assert len(set(df.columns.tolist())&set(adata.obs.columns.tolist())) == 0    \n",
    "    \n",
    "    # merge in first; this way we have access to the genotypes that should\n",
    "    # be in each well\n",
    "    adata.obs = adata.obs.merge(df,\n",
    "                                how='left', \n",
    "                                left_index=True,\n",
    "                                right_index=True)\n",
    "    \n",
    "    # assign genotype for multiplexed wells\n",
    "    df = adata.obs.copy(deep=True) \n",
    "    df = assign_demux_genotype(df)   \n",
    "    \n",
    "    # merge in w/ adata and replace old values in \"Genotype\" \n",
    "    # column for multiplexed wells with the klue results\n",
    "    adata.obs = adata.obs.merge(df, \n",
    "                                how='left',\n",
    "                                left_index=True,\n",
    "                                right_index=True)\n",
    "    inds = adata.obs.loc[adata.obs.well_type=='Multiplexed'].index\n",
    "    adata.obs.Genotype = adata.obs.Genotype.astype('str')\n",
    "    adata.obs.loc[inds, 'Genotype'] = adata.obs.loc[inds, 'new_genotype']\n",
    "    adata.obs.drop('new_genotype', axis=1, inplace=True)\n",
    "    \n",
    "    adata.write(ofile)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "cacc3d4d-6ebd-4374-a700-b49e87d37098",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_kallisto_klue(f, genotypes, ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "51481f3d-d091-4b62-8172-8b2496eae0d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata = sc.read(f)\n",
    "# df = pd.read_csv(genotypes, sep='\\t')\n",
    "# df.set_index('cellID', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "551edd9f-3497-48f7-8c2e-9c471693303c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assert len(set(df.columns.tolist())&set(adata.obs.columns.tolist())) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "aaa005b4-766b-4d99-8aad-db9daaa68da9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # merge in first; this way we have access to the genotypes that should\n",
    "# # be in each well\n",
    "# adata.obs = adata.obs.merge(df,\n",
    "#                             how='left', \n",
    "#                             left_index=True,\n",
    "#                             right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "f4eadf09-f54d-4f99-b395-c750997a4a45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_genotypes():\n",
    "#     g = ['WSBJ','NZOJ',\n",
    "#          'B6J','NODJ','129S1J',\n",
    "#          'CASTJ','AJ','PWKJ']\n",
    "#     return g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "7b34443b-d03a-418f-803e-9d0bd12a3399",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = adata.obs.copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "efc7d234-c11b-418e-a89c-683d212232d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def assign_demux_genotype(df):\n",
    "#     \"\"\"\n",
    "#     Parameters:\n",
    "#         df (pandas DataFrame): DF of obs table for each cell w/ \n",
    "#             klue counts for each genotype and multiplexed genotype\n",
    "#             columns\n",
    "#     \"\"\"\n",
    "#     genotype_cols = get_genotypes()\n",
    "    \n",
    "#     # restrict to nuclei w/ genetic multiplexing\n",
    "#     df = df.loc[df.well_type=='Multiplexed'].copy(deep=True)\n",
    "    \n",
    "#     # fill nans once again\n",
    "#     df[genotype_cols] = df[genotype_cols].fillna(0)\n",
    "    \n",
    "#     # loop through each multiplexed genotype combo\n",
    "#     # use those genotypes to determine which, \n",
    "#     # between the two, has the highest counts\n",
    "#     keep_cols = ['mult_genotype',\n",
    "#                  'mult_genotype_1',\n",
    "#                  'mult_genotype_2']+genotype_cols\n",
    "#     df = df[keep_cols]\n",
    "#     temp2 = pd.DataFrame()\n",
    "#     for g in df.mult_genotype.unique().tolist():\n",
    "#         # print(g)\n",
    "#         temp = df.loc[df.mult_genotype==g].copy(deep=True)\n",
    "\n",
    "#         g1 = temp.mult_genotype_1.unique().tolist()\n",
    "#         assert len(g1) == 1\n",
    "#         g1 = g1[0]\n",
    "\n",
    "#         g2 = temp.mult_genotype_2.unique().tolist()\n",
    "#         assert len(g2) == 1\n",
    "#         g2 = g2[0]\n",
    "    \n",
    "#         # find the best match and report ties if the \n",
    "#         # values are the same\n",
    "#         temp['new_genotype'] = temp[[g1,g2]].idxmax(axis=1)\n",
    "#         temp.loc[temp[g1]==temp[g2], 'new_genotype'] = 'tie'\n",
    "\n",
    "#         temp2 = pd.concat([temp2, temp], axis=0)\n",
    "\n",
    "#     df = df.merge(temp2['new_genotype'], how='left',\n",
    "#                   left_index=True, right_index=True)\n",
    "#     df = df[['new_genotype']]\n",
    "    \n",
    "#     assert len(df.loc[df.new_genotype.isnull()].index) == 0\n",
    "\n",
    "#     return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "5e3920f9-1661-4273-914b-fd3a43244aa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = assign_demux_genotype(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "51760d1a-53fc-4ddc-8fba-02cb4680a3f1",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'new_genotype'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3360\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3361\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'new_genotype'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-227-c0c8e96b16e6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0minds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwell_type\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'Multiplexed'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGenotype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGenotype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'str'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0minds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'Genotype'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0minds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'new_genotype'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'new_genotype'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    923\u001b[0m                 \u001b[0;32mwith\u001b[0m \u001b[0msuppress\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mKeyError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mIndexError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    924\u001b[0m                     \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtakeable\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_takeable\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 925\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    926\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    927\u001b[0m             \u001b[0;31m# we by definition only have the 0th axis\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_tuple\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m   1098\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_getitem_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1099\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0msuppress\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mIndexingError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1100\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_lowerdim\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1101\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1102\u001b[0m         \u001b[0;31m# no multi-index, so validate all of the indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_lowerdim\u001b[0;34m(self, tup)\u001b[0m\n\u001b[1;32m    836\u001b[0m                 \u001b[0;31m# We don't need to check for tuples here because those are\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    837\u001b[0m                 \u001b[0;31m#  caught by the _is_nested_tuple_indexer check above.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 838\u001b[0;31m                 \u001b[0msection\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    839\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    840\u001b[0m                 \u001b[0;31m# We should never have a scalar section here, because\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1162\u001b[0m         \u001b[0;31m# fall thru to straight lookup\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1163\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_key\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1164\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_label\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1165\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1166\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get_slice_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mslice_obj\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mslice\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_get_label\u001b[0;34m(self, label, axis)\u001b[0m\n\u001b[1;32m   1111\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get_label\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1112\u001b[0m         \u001b[0;31m# GH#5667 this will fail if the label is not present in the axis.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1113\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mxs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1114\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1115\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_handle_lowerdim_multi_index_axis0\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtup\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mxs\u001b[0;34m(self, key, axis, level, drop_level)\u001b[0m\n\u001b[1;32m   3759\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3760\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mdrop_level\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3761\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3762\u001b[0m             \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3763\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3456\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3457\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3458\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3459\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3460\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3361\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3362\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3363\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3364\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3365\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhasnans\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'new_genotype'"
     ]
    }
   ],
   "source": [
    "# merge in w/ adata and replace old values in \"Genotype\" \n",
    "# column for multiplexed wells with the klue results\n",
    "adata.obs = adata.obs.merge(df, \n",
    "                            how='left',\n",
    "                            left_index=True,\n",
    "                            right_index=True)\n",
    "inds = adata.obs.loc[adata.obs.well_type=='Multiplexed'].index\n",
    "adata.obs.Genotype = adata.obs.Genotype.astype('str')\n",
    "adata.obs.loc[inds, 'Genotype'] = adata.obs.loc[inds, 'new_genotype']\n",
    "adata.obs.drop('new_genotype', axis=1, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "ffb6fa3b-c62a-4a3f-b81e-2b9b4891c256",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bc</th>\n",
       "      <th>bc1_sequence</th>\n",
       "      <th>bc2_sequence</th>\n",
       "      <th>bc3_sequence</th>\n",
       "      <th>subpool</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>bc1_well</th>\n",
       "      <th>bc2_well</th>\n",
       "      <th>bc3_well</th>\n",
       "      <th>Mouse_Tissue_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>mult_genotype</th>\n",
       "      <th>WSBJ</th>\n",
       "      <th>NZOJ</th>\n",
       "      <th>B6J</th>\n",
       "      <th>NODJ</th>\n",
       "      <th>129S1J</th>\n",
       "      <th>CASTJ</th>\n",
       "      <th>AJ</th>\n",
       "      <th>PWKJ</th>\n",
       "      <th>new_genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [bc, bc1_sequence, bc2_sequence, bc3_sequence, subpool, n_counts, bc1_well, bc2_well, bc3_well, Mouse_Tissue_ID, alias_tissue1, alias_tissue2, alias_tissue3, alias_tissue4, plate, Protocol, Chemistry, well_type, Row, Column, Genotype, Notes, Multiplexed_sample1, Multiplexed_sample2, Tissue, Tissue_ID, Tissue1_ontology_id, Tissue2_ontology_id, Sex, Replicate, DOB, Age_weeks, Age_days, Body_weight_g, Estrus_cycle, Dissection_date, Dissection_time, Tissue_weight_mg, mult_genotype_1, mult_genotype_2, mult_genotype, WSBJ, NZOJ, B6J, NODJ, 129S1J, CASTJ, AJ, PWKJ, new_genotype]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 50 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.loc[adata.obs.Genotype.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "ed290e1a-5d93-4fe0-ac88-3f32f5e09781",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "bc                       object\n",
       "bc1_sequence           category\n",
       "bc2_sequence           category\n",
       "bc3_sequence           category\n",
       "subpool                category\n",
       "n_counts                float64\n",
       "bc1_well               category\n",
       "bc2_well               category\n",
       "bc3_well               category\n",
       "Mouse_Tissue_ID        category\n",
       "alias_tissue1          category\n",
       "alias_tissue2          category\n",
       "alias_tissue3          category\n",
       "alias_tissue4          category\n",
       "plate                  category\n",
       "Protocol               category\n",
       "Chemistry              category\n",
       "well_type              category\n",
       "Row                    category\n",
       "Column                    int64\n",
       "Genotype               category\n",
       "Notes                  category\n",
       "Multiplexed_sample1    category\n",
       "Multiplexed_sample2    category\n",
       "Tissue                 category\n",
       "Tissue_ID                 int64\n",
       "Tissue1_ontology_id    category\n",
       "Tissue2_ontology_id    category\n",
       "Sex                    category\n",
       "Replicate                 int64\n",
       "DOB                    category\n",
       "Age_weeks                 int64\n",
       "Age_days                float64\n",
       "Body_weight_g           float64\n",
       "Estrus_cycle           category\n",
       "Dissection_date        category\n",
       "Dissection_time        category\n",
       "Tissue_weight_mg        float64\n",
       "mult_genotype_1        category\n",
       "mult_genotype_2        category\n",
       "mult_genotype          category\n",
       "WSBJ                    float64\n",
       "NZOJ                    float64\n",
       "B6J                     float64\n",
       "NODJ                    float64\n",
       "129S1J                  float64\n",
       "CASTJ                   float64\n",
       "AJ                      float64\n",
       "PWKJ                    float64\n",
       "new_genotype             object\n",
       "dtype: object"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e5ae959-447e-4cb1-bf84-8928592dffbd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07fbfb3e-a2bf-4895-bf8c-45900733a7c2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "1b4dfcfa-d4c0-4647-bebf-52ecf6d5323b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WSBJ_NZOJ\n",
      "AJ_PWKJ\n",
      "129S1J_CASTJ\n",
      "B6J_NODJ\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/envs/base_clone/lib/python3.7/site-packages/ipykernel_launcher.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/fairliereese/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_column(loc, value, pi)\n"
     ]
    }
   ],
   "source": [
    "# genotype_cols = get_genotypes()\n",
    "# keep_cols = ['mult_genotype',\n",
    "#              'mult_genotype_1',\n",
    "#              'mult_genotype_2']+genotype_cols\n",
    "# df = df[keep_cols]\n",
    "# temp2 = pd.DataFrame()\n",
    "# for g in df.mult_genotype.unique().tolist():\n",
    "#     print(g)\n",
    "#     temp = df.loc[df.mult_genotype==g]\n",
    "    \n",
    "#     g1 = temp.mult_genotype_1.unique().tolist()\n",
    "#     assert len(g1) == 1\n",
    "#     g1 = g1[0]\n",
    "    \n",
    "#     g2 = temp.mult_genotype_2.unique().tolist()\n",
    "#     assert len(g2) == 1\n",
    "#     g2 = g2[0]\n",
    "    \n",
    "#     temp['new_genotype'] = temp[[g1,g2]].idxmax(axis=1)\n",
    "#     temp.loc[temp[g1]==temp[g2], 'new_genotype'] = 'tie'\n",
    "    \n",
    "#     temp2 = pd.concat([temp2, temp], axis=0)\n",
    "    \n",
    "# df = df.merge(temp2['new_genotype'], how='left',\n",
    "#               left_index=True, right_index=True)\n",
    "# # # temp\n",
    "# # print(g)\n",
    "# # print(g1)\n",
    "\n",
    "# # print(g2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "7a0f308d-5362-40de-926a-ddcc3f6467af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mult_genotype</th>\n",
       "      <th>mult_genotype_1</th>\n",
       "      <th>mult_genotype_2</th>\n",
       "      <th>WSBJ</th>\n",
       "      <th>NZOJ</th>\n",
       "      <th>B6J</th>\n",
       "      <th>NODJ</th>\n",
       "      <th>129S1J</th>\n",
       "      <th>CASTJ</th>\n",
       "      <th>AJ</th>\n",
       "      <th>PWKJ</th>\n",
       "      <th>new_genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G10_A2_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H11_A2_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D11_A2_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>AJ_PWKJ</td>\n",
       "      <td>AJ</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>529.0</td>\n",
       "      <td>PWKJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C9_B3_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>AJ_PWKJ</td>\n",
       "      <td>AJ</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C12_B3_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>AJ_PWKJ</td>\n",
       "      <td>AJ</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                mult_genotype mult_genotype_1 mult_genotype_2  \\\n",
       "cellID                                                                          \n",
       "G10_A2_A2_Sublibrary_2_igvf_010     WSBJ_NZOJ            WSBJ            NZOJ   \n",
       "H11_A2_A2_Sublibrary_2_igvf_010     WSBJ_NZOJ            WSBJ            NZOJ   \n",
       "D11_A2_A2_Sublibrary_2_igvf_010       AJ_PWKJ              AJ            PWKJ   \n",
       "C9_B3_A2_Sublibrary_2_igvf_010        AJ_PWKJ              AJ            PWKJ   \n",
       "C12_B3_A2_Sublibrary_2_igvf_010       AJ_PWKJ              AJ            PWKJ   \n",
       "\n",
       "                                 WSBJ  NZOJ  B6J  NODJ  129S1J  CASTJ    AJ  \\\n",
       "cellID                                                                        \n",
       "G10_A2_A2_Sublibrary_2_igvf_010   NaN   NaN  NaN   NaN     NaN    NaN   NaN   \n",
       "H11_A2_A2_Sublibrary_2_igvf_010   NaN   NaN  NaN   NaN     NaN    NaN   NaN   \n",
       "D11_A2_A2_Sublibrary_2_igvf_010   0.0   0.0  0.0   0.0     0.0    0.0  20.0   \n",
       "C9_B3_A2_Sublibrary_2_igvf_010    NaN   NaN  NaN   NaN     NaN    NaN   NaN   \n",
       "C12_B3_A2_Sublibrary_2_igvf_010   NaN   NaN  NaN   NaN     NaN    NaN   NaN   \n",
       "\n",
       "                                  PWKJ new_genotype  \n",
       "cellID                                               \n",
       "G10_A2_A2_Sublibrary_2_igvf_010    NaN          NaN  \n",
       "H11_A2_A2_Sublibrary_2_igvf_010    NaN          NaN  \n",
       "D11_A2_A2_Sublibrary_2_igvf_010  529.0         PWKJ  \n",
       "C9_B3_A2_Sublibrary_2_igvf_010     NaN          NaN  \n",
       "C12_B3_A2_Sublibrary_2_igvf_010    NaN          NaN  "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "6c6730ec-c5fc-425a-9819-fb80b64b5331",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mult_genotype</th>\n",
       "      <th>mult_genotype_1</th>\n",
       "      <th>mult_genotype_2</th>\n",
       "      <th>WSBJ</th>\n",
       "      <th>NZOJ</th>\n",
       "      <th>B6J</th>\n",
       "      <th>NODJ</th>\n",
       "      <th>129S1J</th>\n",
       "      <th>CASTJ</th>\n",
       "      <th>AJ</th>\n",
       "      <th>PWKJ</th>\n",
       "      <th>new_genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>D11_A2_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>AJ_PWKJ</td>\n",
       "      <td>AJ</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>529.0</td>\n",
       "      <td>PWKJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F12_F10_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>129S1J_CASTJ</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>602.0</td>\n",
       "      <td>283.0</td>\n",
       "      <td>295.0</td>\n",
       "      <td>CASTJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F11_F12_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>129S1J_CASTJ</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>CASTJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C12_C9_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>AJ_PWKJ</td>\n",
       "      <td>AJ</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>500.0</td>\n",
       "      <td>602.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1130.0</td>\n",
       "      <td>PWKJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F12_D1_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>129S1J_CASTJ</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>538.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CASTJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D11_E10_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>AJ_PWKJ</td>\n",
       "      <td>AJ</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>568.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>766.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>AJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F9_F1_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>129S1J_CASTJ</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>555.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>517.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>129S1J</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F10_F3_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>129S1J_CASTJ</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>881.0</td>\n",
       "      <td>372.0</td>\n",
       "      <td>407.0</td>\n",
       "      <td>CASTJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C10_F5_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>AJ_PWKJ</td>\n",
       "      <td>AJ</td>\n",
       "      <td>PWKJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>257.0</td>\n",
       "      <td>277.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>536.0</td>\n",
       "      <td>PWKJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E10_F5_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>129S1J_CASTJ</td>\n",
       "      <td>129S1J</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>611.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>553.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>129S1J</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3681 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 mult_genotype mult_genotype_1  \\\n",
       "cellID                                                           \n",
       "D11_A2_A2_Sublibrary_2_igvf_010        AJ_PWKJ              AJ   \n",
       "F12_F10_A2_Sublibrary_2_igvf_010  129S1J_CASTJ          129S1J   \n",
       "F11_F12_A2_Sublibrary_2_igvf_010  129S1J_CASTJ          129S1J   \n",
       "C12_C9_A2_Sublibrary_2_igvf_010        AJ_PWKJ              AJ   \n",
       "F12_D1_A2_Sublibrary_2_igvf_010   129S1J_CASTJ          129S1J   \n",
       "...                                        ...             ...   \n",
       "D11_E10_F8_Sublibrary_2_igvf_010       AJ_PWKJ              AJ   \n",
       "F9_F1_F8_Sublibrary_2_igvf_010    129S1J_CASTJ          129S1J   \n",
       "F10_F3_F8_Sublibrary_2_igvf_010   129S1J_CASTJ          129S1J   \n",
       "C10_F5_F8_Sublibrary_2_igvf_010        AJ_PWKJ              AJ   \n",
       "E10_F5_F8_Sublibrary_2_igvf_010   129S1J_CASTJ          129S1J   \n",
       "\n",
       "                                 mult_genotype_2  WSBJ  NZOJ  B6J  NODJ  \\\n",
       "cellID                                                                    \n",
       "D11_A2_A2_Sublibrary_2_igvf_010             PWKJ   0.0   0.0  0.0   0.0   \n",
       "F12_F10_A2_Sublibrary_2_igvf_010           CASTJ   0.0   0.0  0.0   0.0   \n",
       "F11_F12_A2_Sublibrary_2_igvf_010           CASTJ   0.0   0.0  0.0   0.0   \n",
       "C12_C9_A2_Sublibrary_2_igvf_010             PWKJ   0.0   0.0  0.0   0.0   \n",
       "F12_D1_A2_Sublibrary_2_igvf_010            CASTJ   0.0   0.0  0.0   0.0   \n",
       "...                                          ...   ...   ...  ...   ...   \n",
       "D11_E10_F8_Sublibrary_2_igvf_010            PWKJ   0.0   0.0  0.0   0.0   \n",
       "F9_F1_F8_Sublibrary_2_igvf_010             CASTJ   0.0   0.0  0.0   0.0   \n",
       "F10_F3_F8_Sublibrary_2_igvf_010            CASTJ   0.0   0.0  0.0   0.0   \n",
       "C10_F5_F8_Sublibrary_2_igvf_010             PWKJ   0.0   0.0  0.0   0.0   \n",
       "E10_F5_F8_Sublibrary_2_igvf_010            CASTJ   0.0   0.0  0.0   0.0   \n",
       "\n",
       "                                  129S1J  CASTJ     AJ    PWKJ new_genotype  \n",
       "cellID                                                                       \n",
       "D11_A2_A2_Sublibrary_2_igvf_010      0.0    0.0   20.0   529.0         PWKJ  \n",
       "F12_F10_A2_Sublibrary_2_igvf_010    16.0  602.0  283.0   295.0        CASTJ  \n",
       "F11_F12_A2_Sublibrary_2_igvf_010   158.0  388.0  300.0   202.0        CASTJ  \n",
       "C12_C9_A2_Sublibrary_2_igvf_010    500.0  602.0   30.0  1130.0         PWKJ  \n",
       "F12_D1_A2_Sublibrary_2_igvf_010     19.0  538.0    0.0     0.0        CASTJ  \n",
       "...                                  ...    ...    ...     ...          ...  \n",
       "D11_E10_F8_Sublibrary_2_igvf_010   568.0   97.0  766.0    15.0           AJ  \n",
       "F9_F1_F8_Sublibrary_2_igvf_010     555.0    9.0  517.0    61.0       129S1J  \n",
       "F10_F3_F8_Sublibrary_2_igvf_010     19.0  881.0  372.0   407.0        CASTJ  \n",
       "C10_F5_F8_Sublibrary_2_igvf_010    257.0  277.0   85.0   536.0         PWKJ  \n",
       "E10_F5_F8_Sublibrary_2_igvf_010    611.0    6.0  553.0    77.0       129S1J  \n",
       "\n",
       "[3681 rows x 12 columns]"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[(df.new_genotype.notnull())&(df.new_genotype!='tie')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "65f08175-5ee0-4b16-a587-750df917c8d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2995"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.loc[df.new_genotype=='tie'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7391495-e22e-485d-89d0-92f32fb3ec21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(df.loc[(df.new_genotype=='tie')&(df.])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "1d8a9de1-9485-4f00-8e3a-004134462dde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mult_genotype</th>\n",
       "      <th>mult_genotype_1</th>\n",
       "      <th>mult_genotype_2</th>\n",
       "      <th>WSBJ</th>\n",
       "      <th>NZOJ</th>\n",
       "      <th>B6J</th>\n",
       "      <th>NODJ</th>\n",
       "      <th>129S1J</th>\n",
       "      <th>CASTJ</th>\n",
       "      <th>AJ</th>\n",
       "      <th>PWKJ</th>\n",
       "      <th>new_genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H10_G4_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>494.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>521.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H10_G9_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>519.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>589.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9_A7_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>461.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G10_C11_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>493.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H9_A11_A2_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>475.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G10_D12_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H9_D12_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>504.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H11_E4_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>589.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G10_E10_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>432.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>487.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G9_F3_F8_Sublibrary_2_igvf_010</th>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>446.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>536.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>tie</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1898 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 mult_genotype mult_genotype_1  \\\n",
       "cellID                                                           \n",
       "H10_G4_A2_Sublibrary_2_igvf_010      WSBJ_NZOJ            WSBJ   \n",
       "H10_G9_A2_Sublibrary_2_igvf_010      WSBJ_NZOJ            WSBJ   \n",
       "G9_A7_A2_Sublibrary_2_igvf_010       WSBJ_NZOJ            WSBJ   \n",
       "G10_C11_A2_Sublibrary_2_igvf_010     WSBJ_NZOJ            WSBJ   \n",
       "H9_A11_A2_Sublibrary_2_igvf_010      WSBJ_NZOJ            WSBJ   \n",
       "...                                        ...             ...   \n",
       "G10_D12_F8_Sublibrary_2_igvf_010     WSBJ_NZOJ            WSBJ   \n",
       "H9_D12_F8_Sublibrary_2_igvf_010      WSBJ_NZOJ            WSBJ   \n",
       "H11_E4_F8_Sublibrary_2_igvf_010      WSBJ_NZOJ            WSBJ   \n",
       "G10_E10_F8_Sublibrary_2_igvf_010     WSBJ_NZOJ            WSBJ   \n",
       "G9_F3_F8_Sublibrary_2_igvf_010       WSBJ_NZOJ            WSBJ   \n",
       "\n",
       "                                 mult_genotype_2  WSBJ  NZOJ  B6J  NODJ  \\\n",
       "cellID                                                                    \n",
       "H10_G4_A2_Sublibrary_2_igvf_010             NZOJ   0.0   0.0  0.0   0.0   \n",
       "H10_G9_A2_Sublibrary_2_igvf_010             NZOJ   0.0   0.0  0.0   0.0   \n",
       "G9_A7_A2_Sublibrary_2_igvf_010              NZOJ   0.0   0.0  0.0   0.0   \n",
       "G10_C11_A2_Sublibrary_2_igvf_010            NZOJ   0.0   0.0  0.0   0.0   \n",
       "H9_A11_A2_Sublibrary_2_igvf_010             NZOJ   0.0   0.0  0.0   0.0   \n",
       "...                                          ...   ...   ...  ...   ...   \n",
       "G10_D12_F8_Sublibrary_2_igvf_010            NZOJ   0.0   0.0  0.0   0.0   \n",
       "H9_D12_F8_Sublibrary_2_igvf_010             NZOJ   0.0   0.0  0.0   0.0   \n",
       "H11_E4_F8_Sublibrary_2_igvf_010             NZOJ   0.0   0.0  0.0   0.0   \n",
       "G10_E10_F8_Sublibrary_2_igvf_010            NZOJ   0.0   0.0  0.0   0.0   \n",
       "G9_F3_F8_Sublibrary_2_igvf_010              NZOJ   0.0   0.0  0.0   0.0   \n",
       "\n",
       "                                  129S1J  CASTJ     AJ   PWKJ new_genotype  \n",
       "cellID                                                                      \n",
       "H10_G4_A2_Sublibrary_2_igvf_010    494.0   77.0  521.0  109.0          tie  \n",
       "H10_G9_A2_Sublibrary_2_igvf_010    519.0   99.0  589.0   88.0          tie  \n",
       "G9_A7_A2_Sublibrary_2_igvf_010       0.0    0.0  461.0   57.0          tie  \n",
       "G10_C11_A2_Sublibrary_2_igvf_010     0.0    0.0  493.0   62.0          tie  \n",
       "H9_A11_A2_Sublibrary_2_igvf_010      0.0    0.0  475.0   93.0          tie  \n",
       "...                                  ...    ...    ...    ...          ...  \n",
       "G10_D12_F8_Sublibrary_2_igvf_010     0.0    0.0  511.0   54.0          tie  \n",
       "H9_D12_F8_Sublibrary_2_igvf_010      0.0    0.0  504.0   83.0          tie  \n",
       "H11_E4_F8_Sublibrary_2_igvf_010      0.0    0.0  589.0   47.0          tie  \n",
       "G10_E10_F8_Sublibrary_2_igvf_010   432.0   89.0  487.0   74.0          tie  \n",
       "G9_F3_F8_Sublibrary_2_igvf_010     446.0   97.0  536.0   84.0          tie  \n",
       "\n",
       "[1898 rows x 12 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.WSBJ==temp.NZOJ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "5aed8074-a25d-47e6-9b96-be1d81bae3e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mult_genotype</th>\n",
       "      <th>mult_genotype_1</th>\n",
       "      <th>mult_genotype_2</th>\n",
       "      <th>WSBJ</th>\n",
       "      <th>NZOJ</th>\n",
       "      <th>B6J</th>\n",
       "      <th>NODJ</th>\n",
       "      <th>129S1J</th>\n",
       "      <th>CASTJ</th>\n",
       "      <th>AJ</th>\n",
       "      <th>PWKJ</th>\n",
       "      <th>new_genotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [mult_genotype, mult_genotype_1, mult_genotype_2, WSBJ, NZOJ, B6J, NODJ, 129S1J, CASTJ, AJ, PWKJ, new_genotype]\n",
       "Index: []"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.new_genotype == 'WSB']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4c81d87-eeb3-4fc3-acb0-4ffc066fc4d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82e8ea2f-8b8b-4ca0-bc0c-f58fb6b23fba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "20c4f30a-50f0-4ebf-83f2-1111cb00f1be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['WSBJ/NZOJ', 'AJ', 'AJ/PWKJ', 'NODJ', 'B6J', ..., 'NZOJ', '129S1J/CASTJ', 'B6J/NODJ', 'WSBJ', 'CASTJ']\n",
       "Length: 12\n",
       "Categories (12, object): ['129S1J', '129S1J/CASTJ', 'AJ', 'AJ/PWKJ', ..., 'NZOJ', 'PWKJ', 'WSBJ', 'WSBJ/NZOJ']"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.Genotype.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3bedf2b-7fad-47df-9fb5-1404f82ad8f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fb3c007-4208-4f3c-a365-0a9a30f74dee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c11c3d1a-d86e-4c9f-affb-1894fd2c98ca",
   "metadata": {},
   "source": [
    "## 231014 filter / do some other reformatting for klue output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "e51ab42e-7765-4be5-baf2-347dda540d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "094ca9a8-80b1-463f-ae0c-9792db921165",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = ['/Users/fairliereese/mortazavi_lab/bin/igvf_pipeline/klue/igvf_010/Sublibrary_2/129S1J_CASTJ/filtered/adata.h5ad', \n",
    "         '/Users/fairliereese/mortazavi_lab/bin/igvf_pipeline/klue/igvf_010/Sublibrary_2/AJ_PWKJ/filtered/adata.h5ad']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ac63d3f-79a6-4a62-a5c6-c019a06961ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genotype_counts(files, ofile):\n",
    "    \"\"\"\n",
    "    Given a list of klue anndata objects, merge \n",
    "    the genotype counts together for each cell and output\n",
    "    as a tsv\n",
    "    \"\"\"\n",
    "    for i, f in enumerate(files):\n",
    "        adata = sc.read(f)\n",
    "        if i == 0:\n",
    "            df = adata.to_df()\n",
    "        else:\n",
    "            df = df.merge(adata.to_df(),\n",
    "                          how='outer',\n",
    "                          left_index=True,\n",
    "                          right_index=True)\n",
    "    df.to_csv(ofile, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "ef248cd2-543b-4159-91bc-4401fc75a245",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, f in enumerate(files):\n",
    "    adata = sc.read(f)\n",
    "    if i == 0:\n",
    "        df = adata.to_df()\n",
    "    else:\n",
    "        df = df.merge(adata.to_df(),\n",
    "                      how='outer',\n",
    "                      left_index=True,\n",
    "                      right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "5eca475b-d6eb-4c1c-aacd-f28f4475293d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>genotype</th>\n",
       "      <th>129S1J</th>\n",
       "      <th>CASTJ</th>\n",
       "      <th>AJ</th>\n",
       "      <th>PWKJ</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A10_A12_F9_Sublibrary_2_igvf_010</th>\n",
       "      <td>512.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>643.0</td>\n",
       "      <td>87.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A10_A1_D9_Sublibrary_2_igvf_010</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>430.0</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A10_A1_E7_Sublibrary_2_igvf_010</th>\n",
       "      <td>716.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>1077.0</td>\n",
       "      <td>182.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A10_A4_C3_Sublibrary_2_igvf_010</th>\n",
       "      <td>477.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>564.0</td>\n",
       "      <td>75.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A10_A4_G3_Sublibrary_2_igvf_010</th>\n",
       "      <td>438.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>583.0</td>\n",
       "      <td>113.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "genotype                          129S1J  CASTJ      AJ   PWKJ\n",
       "cellID                                                        \n",
       "A10_A12_F9_Sublibrary_2_igvf_010   512.0   86.0   643.0   87.0\n",
       "A10_A1_D9_Sublibrary_2_igvf_010      NaN    NaN   430.0   80.0\n",
       "A10_A1_E7_Sublibrary_2_igvf_010    716.0  115.0  1077.0  182.0\n",
       "A10_A4_C3_Sublibrary_2_igvf_010    477.0   90.0   564.0   75.0\n",
       "A10_A4_G3_Sublibrary_2_igvf_010    438.0   90.0   583.0  113.0"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "bed33860-d327-4136-b3f7-b66f12745b6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffa09bbc-4c81-47fd-aa7a-dbdd1eb5e31f",
   "metadata": {},
   "source": [
    "## 231014 How to ask for all klue outputs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c6d4bb5c-a8cb-4b1b-afff-1aa34eb00fc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_tsv = '../configs/test_4.tsv'\n",
    "sample_csv = '../configs/sample_metadata.csv'\n",
    "kit = 'WT_mega'\n",
    "chemistry = 'v2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "78c1f868-3e37-46f6-841b-27316bbf35c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in config / analysis spec\n",
    "df = parse_config(config_tsv)\n",
    "bc_df = get_bc1_matches(kit, chemistry)\n",
    "sample_df = parse_sample_df(sample_csv)\n",
    "\n",
    "mult_genotype_1s = sample_df.loc[sample_df.mult_genotype_1.notnull(), 'mult_genotype_1'].unique().tolist()\n",
    "mult_genotype_2s = sample_df.loc[sample_df.mult_genotype_2.notnull(), 'mult_genotype_2'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "88a425b8-442d-495a-a7f0-da73ce24775a",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = df.merge(sample_df, on='plate', how='inner')\n",
    "temp = temp.loc[temp.mult_genotype.notnull()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3761b301-f37a-4b22-a5e3-f6c3c51c97c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fastq</th>\n",
       "      <th>subpool</th>\n",
       "      <th>plate</th>\n",
       "      <th>lane</th>\n",
       "      <th>run</th>\n",
       "      <th>path</th>\n",
       "      <th>path2</th>\n",
       "      <th>r2_fastq</th>\n",
       "      <th>Mouse_Tissue_ID</th>\n",
       "      <th>alias_tissue1</th>\n",
       "      <th>...</th>\n",
       "      <th>Age_weeks</th>\n",
       "      <th>Age_days</th>\n",
       "      <th>Body_weight_g</th>\n",
       "      <th>Estrus_cycle</th>\n",
       "      <th>Dissection_date</th>\n",
       "      <th>Dissection_time</th>\n",
       "      <th>Tissue_weight_mg</th>\n",
       "      <th>mult_genotype_1</th>\n",
       "      <th>mult_genotype_2</th>\n",
       "      <th>mult_genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>Sublibrary_2</td>\n",
       "      <td>igvf_010</td>\n",
       "      <td>L001</td>\n",
       "      <td>1</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>AB_M1_16</td>\n",
       "      <td>ali-mortazavi:017_B6J_10M_16_UBERON_0001388</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>Sublibrary_2</td>\n",
       "      <td>igvf_010</td>\n",
       "      <td>L001</td>\n",
       "      <td>1</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>AB_F2_16</td>\n",
       "      <td>ali-mortazavi:018_B6J_10F_16_UBERON_0001388</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>Sublibrary_2</td>\n",
       "      <td>igvf_010</td>\n",
       "      <td>L001</td>\n",
       "      <td>1</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>AB_M2_16</td>\n",
       "      <td>ali-mortazavi:019_B6J_10M_16_UBERON_0001388</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>Sublibrary_2</td>\n",
       "      <td>igvf_010</td>\n",
       "      <td>L001</td>\n",
       "      <td>1</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>AB_F1_16</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F_16_UBERON_0001388</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>Sublibrary_2</td>\n",
       "      <td>igvf_010</td>\n",
       "      <td>L001</td>\n",
       "      <td>1</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/</td>\n",
       "      <td>/dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...</td>\n",
       "      <td>AB_M3_16</td>\n",
       "      <td>ali-mortazavi:021_B6J_10M_16_UBERON_0001388</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                fastq       subpool     plate  \\\n",
       "64  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...  Sublibrary_2  igvf_010   \n",
       "65  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...  Sublibrary_2  igvf_010   \n",
       "66  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...  Sublibrary_2  igvf_010   \n",
       "67  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...  Sublibrary_2  igvf_010   \n",
       "68  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...  Sublibrary_2  igvf_010   \n",
       "\n",
       "    lane  run                                 path  \\\n",
       "64  L001    1  /dfs7/samlab/seyedam/IGVF/igvf_010/   \n",
       "65  L001    1  /dfs7/samlab/seyedam/IGVF/igvf_010/   \n",
       "66  L001    1  /dfs7/samlab/seyedam/IGVF/igvf_010/   \n",
       "67  L001    1  /dfs7/samlab/seyedam/IGVF/igvf_010/   \n",
       "68  L001    1  /dfs7/samlab/seyedam/IGVF/igvf_010/   \n",
       "\n",
       "                                        path2  \\\n",
       "64  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/   \n",
       "65  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/   \n",
       "66  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/   \n",
       "67  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/   \n",
       "68  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/   \n",
       "\n",
       "                                             r2_fastq Mouse_Tissue_ID  \\\n",
       "64  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...        AB_M1_16   \n",
       "65  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...        AB_F2_16   \n",
       "66  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...        AB_M2_16   \n",
       "67  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...        AB_F1_16   \n",
       "68  /dfs7/samlab/seyedam/IGVF/igvf_010/nova1/Subli...        AB_M3_16   \n",
       "\n",
       "                                  alias_tissue1  ... Age_weeks Age_days  \\\n",
       "64  ali-mortazavi:017_B6J_10M_16_UBERON_0001388  ...        10      NaN   \n",
       "65  ali-mortazavi:018_B6J_10F_16_UBERON_0001388  ...        10      NaN   \n",
       "66  ali-mortazavi:019_B6J_10M_16_UBERON_0001388  ...        10      NaN   \n",
       "67  ali-mortazavi:016_B6J_10F_16_UBERON_0001388  ...        10      NaN   \n",
       "68  ali-mortazavi:021_B6J_10M_16_UBERON_0001388  ...        10      NaN   \n",
       "\n",
       "   Body_weight_g Estrus_cycle Dissection_date Dissection_time  \\\n",
       "64           NaN          NaN             NaN             NaN   \n",
       "65           NaN          NaN             NaN             NaN   \n",
       "66           NaN          NaN             NaN             NaN   \n",
       "67           NaN          NaN             NaN             NaN   \n",
       "68           NaN          NaN             NaN             NaN   \n",
       "\n",
       "   Tissue_weight_mg mult_genotype_1  mult_genotype_2 mult_genotype  \n",
       "64              NaN             B6J             NODJ      B6J_NODJ  \n",
       "65              NaN             B6J             NODJ      B6J_NODJ  \n",
       "66              NaN             B6J             NODJ      B6J_NODJ  \n",
       "67              NaN             B6J             NODJ      B6J_NODJ  \n",
       "68              NaN             B6J             NODJ      B6J_NODJ  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7413804a-fc76-465a-899b-db10413d56c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ae19bad-8ff9-4d99-aba7-f179c1a6bb50",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a75b13e5-215c-4616-a4a5-4340ba89869f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # def get_subset_mult_genotypes(df, sample_df):\n",
    "# temp = df.merge(sample_df, on='plate', how='inner')\n",
    "# temp = temp.loc[temp.mult_genotype.notnull()]\n",
    "# mult_genotype_1s = temp.mult_genotype_1.tolist()\n",
    "# mult_genotype_2s = temp.mult_genotype_2.tolist()\n",
    "# # return mult_genotype_1s, mult_genotype_2s\n",
    "\n",
    "def get_subset_mult_genotypes(df, sample_df, col):\n",
    "    temp = df.merge(sample_df, on='plate', how='inner')\n",
    "    temp = temp.loc[temp.mult_genotype.notnull()]\n",
    "    entries = temp[col].tolist()\n",
    "    return entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "07b380f5-f243-4764-8997-6f056adcd616",
   "metadata": {},
   "outputs": [],
   "source": [
    "entries = get_subset_mult_genotypes(df, sample_df, 'mult_genotype_1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9f80d503-0468-4fa1-a13c-9555045cf73d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'B6J',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " 'AJ',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " '129S1J',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ',\n",
       " 'WSBJ']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "076ce088-fe08-49f5-8a23-a4abfb0ce36a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2128c11f-c59d-4f36-9b3d-075a243cc7a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e363356f-fb6c-48b2-9fb6-baa855b99ac4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2419856-529a-49c2-8d2e-d4aeb1ea5ee5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "079896f1-0949-481a-a1a1-49a0a337563d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "549f06f4-b293-40a9-ae3e-9a8b6fd2c255",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cc82bb95-1adf-4033-a067-56613cb243f6",
   "metadata": {},
   "source": [
    "## 231014 adding multiplexed genotype as fields to the sample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bbdf5b88-a192-44d2-8bd4-c292c58bff3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_csv = '../configs/sample_metadata.csv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "cbc59b84-709e-4262-bd48-96f48e9e3099",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_sample_df(fname):\n",
    "    df = pd.read_csv(fname)\n",
    "\n",
    "    # add multiplexed genotypes if relevant\n",
    "    g_cols = ['Multiplexed_Genotype_1', 'Multiplexed_Genotype_2']\n",
    "    df[g_cols] = df.Genotype.str.split('/', expand=True)\n",
    "\n",
    "    # adjust single-genotype wells\n",
    "    df.loc[df.well_type=='Single', g_cols] = np.nan\n",
    "\n",
    "    # add a multiplexed genotype column\n",
    "    inds = df.loc[df.well_type=='Multiplexed'].index\n",
    "    df['mult_genotype'] = np.nan\n",
    "    df.loc[inds, 'mult_genotype'] = df.loc[inds, g_cols[0]]+'_'+\\\n",
    "                                   df.loc[inds, g_cols[1]]\n",
    "\n",
    "    # checks\n",
    "    for g in g_cols:\n",
    "        assert len(df.loc[(df.well_type=='Single')&(df[g].notnull())].index) == 0\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "ec6018fe-b0fc-4f2d-a6c4-40bc1239b125",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = parse_sample_df(sample_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c84e9c5b-7a0c-408e-8007-20a542373651",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse_Tissue_ID</th>\n",
       "      <th>alias_tissue1</th>\n",
       "      <th>alias_tissue2</th>\n",
       "      <th>alias_tissue3</th>\n",
       "      <th>alias_tissue4</th>\n",
       "      <th>plate</th>\n",
       "      <th>Protocol</th>\n",
       "      <th>Chemistry</th>\n",
       "      <th>bc1_well</th>\n",
       "      <th>well_type</th>\n",
       "      <th>...</th>\n",
       "      <th>Age_weeks</th>\n",
       "      <th>Age_days</th>\n",
       "      <th>Body_weight_g</th>\n",
       "      <th>Estrus_cycle</th>\n",
       "      <th>Dissection_date</th>\n",
       "      <th>Dissection_time</th>\n",
       "      <th>Tissue_weight_mg</th>\n",
       "      <th>Multiplexed_Genotype_1</th>\n",
       "      <th>Multiplexed_Genotype_2</th>\n",
       "      <th>mult_genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>AB_M1_06</td>\n",
       "      <td>ali-mortazavi:017_B6J_10M_06_UBERON_0000948</td>\n",
       "      <td>ali-mortazavi:067_NODJ_10M_06_UBERON_0000948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>A10</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>AB_F2_06</td>\n",
       "      <td>ali-mortazavi:018_B6J_10F_06_UBERON_0000948</td>\n",
       "      <td>ali-mortazavi:068_NODJ_10F_06_UBERON_0000948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>A11</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>AB_M2_06</td>\n",
       "      <td>ali-mortazavi:019_B6J_10M_06_UBERON_0000948</td>\n",
       "      <td>ali-mortazavi:069_NODJ_10M_06_UBERON_0000948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>A12</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>AB_F1_06</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F_06_UBERON_0000948</td>\n",
       "      <td>ali-mortazavi:066_NODJ_10F_06_UBERON_0000948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>A9</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>AB_M3_06</td>\n",
       "      <td>ali-mortazavi:021_B6J_10M_06_UBERON_0000948</td>\n",
       "      <td>ali-mortazavi:071_NODJ_10M_06_UBERON_0000948</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>B10</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B6J</td>\n",
       "      <td>NODJ</td>\n",
       "      <td>B6J_NODJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>859</th>\n",
       "      <td>GH_F1_03</td>\n",
       "      <td>ali-mortazavi:058_WSBJ_10F_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:092_CASTJ_10F_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:058_WSBJ_10F_03_NTR_0000750</td>\n",
       "      <td>ali-mortazavi:092_CASTJ_10F_03_NTR_0000750</td>\n",
       "      <td>igvf_011</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>G9</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>CASTJ</td>\n",
       "      <td>WSBJ_CASTJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>GH_M3_03</td>\n",
       "      <td>ali-mortazavi:061_WSBJ_10M_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:051_NZOJ_10M_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:061_WSBJ_10M_03_NTR_0000750</td>\n",
       "      <td>ali-mortazavi:051_NZOJ_10M_03_NTR_0000750</td>\n",
       "      <td>igvf_011</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>H10</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>861</th>\n",
       "      <td>GH_F4_03</td>\n",
       "      <td>ali-mortazavi:096_WSBJ_10F_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:052_NZOJ_10F_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:096_WSBJ_10F_03_NTR_0000750</td>\n",
       "      <td>ali-mortazavi:052_NZOJ_10F_03_NTR_0000750</td>\n",
       "      <td>igvf_011</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>H11</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>862</th>\n",
       "      <td>GH_M4_03</td>\n",
       "      <td>ali-mortazavi:063_WSBJ_10M_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:053_NZOJ_10M_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:063_WSBJ_10M_03_NTR_0000750</td>\n",
       "      <td>ali-mortazavi:053_NZOJ_10M_03_NTR_0000750</td>\n",
       "      <td>igvf_011</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>H12</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>GH_F3_03</td>\n",
       "      <td>ali-mortazavi:062_WSBJ_10F_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:050_NZOJ_10F_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:062_WSBJ_10F_03_NTR_0000750</td>\n",
       "      <td>ali-mortazavi:050_NZOJ_10F_03_NTR_0000750</td>\n",
       "      <td>igvf_011</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>H9</td>\n",
       "      <td>Multiplexed</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WSBJ</td>\n",
       "      <td>NZOJ</td>\n",
       "      <td>WSBJ_NZOJ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>288 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Mouse_Tissue_ID                                alias_tissue1  \\\n",
       "64         AB_M1_06  ali-mortazavi:017_B6J_10M_06_UBERON_0000948   \n",
       "65         AB_F2_06  ali-mortazavi:018_B6J_10F_06_UBERON_0000948   \n",
       "66         AB_M2_06  ali-mortazavi:019_B6J_10M_06_UBERON_0000948   \n",
       "67         AB_F1_06  ali-mortazavi:016_B6J_10F_06_UBERON_0000948   \n",
       "68         AB_M3_06  ali-mortazavi:021_B6J_10M_06_UBERON_0000948   \n",
       "..              ...                                          ...   \n",
       "859        GH_F1_03    ali-mortazavi:058_WSBJ_10F_03_NTR_0000646   \n",
       "860        GH_M3_03    ali-mortazavi:061_WSBJ_10M_03_NTR_0000646   \n",
       "861        GH_F4_03    ali-mortazavi:096_WSBJ_10F_03_NTR_0000646   \n",
       "862        GH_M4_03    ali-mortazavi:063_WSBJ_10M_03_NTR_0000646   \n",
       "863        GH_F3_03    ali-mortazavi:062_WSBJ_10F_03_NTR_0000646   \n",
       "\n",
       "                                    alias_tissue2  \\\n",
       "64   ali-mortazavi:067_NODJ_10M_06_UBERON_0000948   \n",
       "65   ali-mortazavi:068_NODJ_10F_06_UBERON_0000948   \n",
       "66   ali-mortazavi:069_NODJ_10M_06_UBERON_0000948   \n",
       "67   ali-mortazavi:066_NODJ_10F_06_UBERON_0000948   \n",
       "68   ali-mortazavi:071_NODJ_10M_06_UBERON_0000948   \n",
       "..                                            ...   \n",
       "859    ali-mortazavi:092_CASTJ_10F_03_NTR_0000646   \n",
       "860     ali-mortazavi:051_NZOJ_10M_03_NTR_0000646   \n",
       "861     ali-mortazavi:052_NZOJ_10F_03_NTR_0000646   \n",
       "862     ali-mortazavi:053_NZOJ_10M_03_NTR_0000646   \n",
       "863     ali-mortazavi:050_NZOJ_10F_03_NTR_0000646   \n",
       "\n",
       "                                 alias_tissue3  \\\n",
       "64                                         NaN   \n",
       "65                                         NaN   \n",
       "66                                         NaN   \n",
       "67                                         NaN   \n",
       "68                                         NaN   \n",
       "..                                         ...   \n",
       "859  ali-mortazavi:058_WSBJ_10F_03_NTR_0000750   \n",
       "860  ali-mortazavi:061_WSBJ_10M_03_NTR_0000750   \n",
       "861  ali-mortazavi:096_WSBJ_10F_03_NTR_0000750   \n",
       "862  ali-mortazavi:063_WSBJ_10M_03_NTR_0000750   \n",
       "863  ali-mortazavi:062_WSBJ_10F_03_NTR_0000750   \n",
       "\n",
       "                                  alias_tissue4     plate       Protocol  \\\n",
       "64                                          NaN  igvf_003  Parse_WT_Mega   \n",
       "65                                          NaN  igvf_003  Parse_WT_Mega   \n",
       "66                                          NaN  igvf_003  Parse_WT_Mega   \n",
       "67                                          NaN  igvf_003  Parse_WT_Mega   \n",
       "68                                          NaN  igvf_003  Parse_WT_Mega   \n",
       "..                                          ...       ...            ...   \n",
       "859  ali-mortazavi:092_CASTJ_10F_03_NTR_0000750  igvf_011  Parse_WT_Mega   \n",
       "860   ali-mortazavi:051_NZOJ_10M_03_NTR_0000750  igvf_011  Parse_WT_Mega   \n",
       "861   ali-mortazavi:052_NZOJ_10F_03_NTR_0000750  igvf_011  Parse_WT_Mega   \n",
       "862   ali-mortazavi:053_NZOJ_10M_03_NTR_0000750  igvf_011  Parse_WT_Mega   \n",
       "863   ali-mortazavi:050_NZOJ_10F_03_NTR_0000750  igvf_011  Parse_WT_Mega   \n",
       "\n",
       "    Chemistry bc1_well    well_type  ... Age_weeks  Age_days Body_weight_g  \\\n",
       "64         v2      A10  Multiplexed  ...        10       NaN           NaN   \n",
       "65         v2      A11  Multiplexed  ...        10       NaN           NaN   \n",
       "66         v2      A12  Multiplexed  ...        10       NaN           NaN   \n",
       "67         v2       A9  Multiplexed  ...        10       NaN           NaN   \n",
       "68         v2      B10  Multiplexed  ...        10       NaN           NaN   \n",
       "..        ...      ...          ...  ...       ...       ...           ...   \n",
       "859        v2       G9  Multiplexed  ...        10       NaN           NaN   \n",
       "860        v2      H10  Multiplexed  ...        10       NaN           NaN   \n",
       "861        v2      H11  Multiplexed  ...        10       NaN           NaN   \n",
       "862        v2      H12  Multiplexed  ...        10       NaN           NaN   \n",
       "863        v2       H9  Multiplexed  ...        10       NaN           NaN   \n",
       "\n",
       "    Estrus_cycle Dissection_date Dissection_time Tissue_weight_mg  \\\n",
       "64           NaN             NaN             NaN              NaN   \n",
       "65           NaN             NaN             NaN              NaN   \n",
       "66           NaN             NaN             NaN              NaN   \n",
       "67           NaN             NaN             NaN              NaN   \n",
       "68           NaN             NaN             NaN              NaN   \n",
       "..           ...             ...             ...              ...   \n",
       "859          NaN             NaN             NaN              NaN   \n",
       "860          NaN             NaN             NaN              NaN   \n",
       "861          NaN             NaN             NaN              NaN   \n",
       "862          NaN             NaN             NaN              NaN   \n",
       "863          NaN             NaN             NaN              NaN   \n",
       "\n",
       "     Multiplexed_Genotype_1 Multiplexed_Genotype_2 mult_genotype  \n",
       "64                      B6J                   NODJ      B6J_NODJ  \n",
       "65                      B6J                   NODJ      B6J_NODJ  \n",
       "66                      B6J                   NODJ      B6J_NODJ  \n",
       "67                      B6J                   NODJ      B6J_NODJ  \n",
       "68                      B6J                   NODJ      B6J_NODJ  \n",
       "..                      ...                    ...           ...  \n",
       "859                    WSBJ                  CASTJ    WSBJ_CASTJ  \n",
       "860                    WSBJ                   NZOJ     WSBJ_NZOJ  \n",
       "861                    WSBJ                   NZOJ     WSBJ_NZOJ  \n",
       "862                    WSBJ                   NZOJ     WSBJ_NZOJ  \n",
       "863                    WSBJ                   NZOJ     WSBJ_NZOJ  \n",
       "\n",
       "[288 rows x 33 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.well_type=='Multiplexed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "f06471e5-8a89-4e5b-8e4e-9a700e255304",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "bd882d6e-bd7d-458c-89a5-305895314573",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse_Tissue_ID</th>\n",
       "      <th>alias_tissue1</th>\n",
       "      <th>alias_tissue2</th>\n",
       "      <th>alias_tissue3</th>\n",
       "      <th>alias_tissue4</th>\n",
       "      <th>plate</th>\n",
       "      <th>Protocol</th>\n",
       "      <th>Chemistry</th>\n",
       "      <th>bc1_well</th>\n",
       "      <th>well_type</th>\n",
       "      <th>...</th>\n",
       "      <th>DOB</th>\n",
       "      <th>Age_weeks</th>\n",
       "      <th>Age_days</th>\n",
       "      <th>Body_weight_g</th>\n",
       "      <th>Estrus_cycle</th>\n",
       "      <th>Dissection_date</th>\n",
       "      <th>Dissection_time</th>\n",
       "      <th>Tissue_weight_mg</th>\n",
       "      <th>Multiplexed_Genotype_1</th>\n",
       "      <th>Multiplexed_Genotype_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Mouse_Tissue_ID, alias_tissue1, alias_tissue2, alias_tissue3, alias_tissue4, plate, Protocol, Chemistry, bc1_well, well_type, Row, Column, Genotype, Notes, Multiplexed_sample1, Multiplexed_sample2, Tissue, Tissue_ID, Tissue1_ontology_id, Tissue2_ontology_id, Sex, Replicate, DOB, Age_weeks, Age_days, Body_weight_g, Estrus_cycle, Dissection_date, Dissection_time, Tissue_weight_mg, Multiplexed_Genotype_1, Multiplexed_Genotype_2]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 32 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g ='Multiplexed_Genotype_1'\n",
    "df.loc[(df.well_type=='Single')&(df[g].notnull())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba3a5f8f-fd4a-49bc-ad0a-02a6acae33bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "21e516c5-c65e-43d8-bf0e-dde0f9f38bad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse_Tissue_ID</th>\n",
       "      <th>alias_tissue1</th>\n",
       "      <th>alias_tissue2</th>\n",
       "      <th>alias_tissue3</th>\n",
       "      <th>alias_tissue4</th>\n",
       "      <th>plate</th>\n",
       "      <th>Protocol</th>\n",
       "      <th>Chemistry</th>\n",
       "      <th>bc1_well</th>\n",
       "      <th>well_type</th>\n",
       "      <th>...</th>\n",
       "      <th>Age_weeks</th>\n",
       "      <th>Age_days</th>\n",
       "      <th>Body_weight_g</th>\n",
       "      <th>Estrus_cycle</th>\n",
       "      <th>Dissection_date</th>\n",
       "      <th>Dissection_time</th>\n",
       "      <th>Tissue_weight_mg</th>\n",
       "      <th>Multiplexed_Genotype_1</th>\n",
       "      <th>Multiplexed_Genotype_2</th>\n",
       "      <th>mult_genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>016_B6J_10F_03</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:016_B6J_10F_03_NTR_0000750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>A1</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>72.0</td>\n",
       "      <td>21.1</td>\n",
       "      <td>Diestrus</td>\n",
       "      <td>10/27/22</td>\n",
       "      <td>9:01 AM</td>\n",
       "      <td>141.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>017_B6J_10M_03</td>\n",
       "      <td>ali-mortazavi:017_B6J_10M_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:017_B6J_10M_03_NTR_0000750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>A2</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>72.0</td>\n",
       "      <td>26.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10/27/22</td>\n",
       "      <td>11:44 AM</td>\n",
       "      <td>141.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>018_B6J_10F_03</td>\n",
       "      <td>ali-mortazavi:018_B6J_10F_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:018_B6J_10F_03_NTR_0000750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>A3</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>72.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Proestrus</td>\n",
       "      <td>10/27/22</td>\n",
       "      <td>9:38 AM</td>\n",
       "      <td>143.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>019_B6J_10M_03</td>\n",
       "      <td>ali-mortazavi:019_B6J_10M_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:019_B6J_10M_03_NTR_0000750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>A4</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>72.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10/27/22</td>\n",
       "      <td>12:12 PM</td>\n",
       "      <td>141.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>020_B6J_10F_03</td>\n",
       "      <td>ali-mortazavi:020_B6J_10F_03_NTR_0000646</td>\n",
       "      <td>ali-mortazavi:020_B6J_10F_03_NTR_0000750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_003</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>A5</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>72.0</td>\n",
       "      <td>18.5</td>\n",
       "      <td>Diestrus</td>\n",
       "      <td>10/27/22</td>\n",
       "      <td>10:14 AM</td>\n",
       "      <td>149.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1051</th>\n",
       "      <td>163_B6NODF1J_10M_20</td>\n",
       "      <td>ali-mortazavi:163_B6NODF1J_10M_20_CL_2000001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_013</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>D12</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7/24/23</td>\n",
       "      <td>1:38 PM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1052</th>\n",
       "      <td>166_B6NODF1J_10F_20</td>\n",
       "      <td>ali-mortazavi:166_B6NODF1J_10F_20_CL_2000001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_013</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>E12</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7/24/23</td>\n",
       "      <td>11:10 AM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053</th>\n",
       "      <td>165_B6NODF1J_10M_20</td>\n",
       "      <td>ali-mortazavi:165_B6NODF1J_10M_20_CL_2000001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_013</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>F12</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7/24/23</td>\n",
       "      <td>1:58 PM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1054</th>\n",
       "      <td>168_B6NODF1J_10F_20</td>\n",
       "      <td>ali-mortazavi:168_B6NODF1J_10F_20_CL_2000001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_013</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>G12</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7/24/23</td>\n",
       "      <td>11:41 AM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1055</th>\n",
       "      <td>167_B6NODF1J_10M_20</td>\n",
       "      <td>ali-mortazavi:167_B6NODF1J_10M_20_CL_2000001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>igvf_013</td>\n",
       "      <td>Parse_WT_Mega</td>\n",
       "      <td>v2</td>\n",
       "      <td>H12</td>\n",
       "      <td>Single</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7/24/23</td>\n",
       "      <td>2:21 PM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>768 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Mouse_Tissue_ID                                 alias_tissue1  \\\n",
       "0          016_B6J_10F_03      ali-mortazavi:016_B6J_10F_03_NTR_0000646   \n",
       "1          017_B6J_10M_03      ali-mortazavi:017_B6J_10M_03_NTR_0000646   \n",
       "2          018_B6J_10F_03      ali-mortazavi:018_B6J_10F_03_NTR_0000646   \n",
       "3          019_B6J_10M_03      ali-mortazavi:019_B6J_10M_03_NTR_0000646   \n",
       "4          020_B6J_10F_03      ali-mortazavi:020_B6J_10F_03_NTR_0000646   \n",
       "...                   ...                                           ...   \n",
       "1051  163_B6NODF1J_10M_20  ali-mortazavi:163_B6NODF1J_10M_20_CL_2000001   \n",
       "1052  166_B6NODF1J_10F_20  ali-mortazavi:166_B6NODF1J_10F_20_CL_2000001   \n",
       "1053  165_B6NODF1J_10M_20  ali-mortazavi:165_B6NODF1J_10M_20_CL_2000001   \n",
       "1054  168_B6NODF1J_10F_20  ali-mortazavi:168_B6NODF1J_10F_20_CL_2000001   \n",
       "1055  167_B6NODF1J_10M_20  ali-mortazavi:167_B6NODF1J_10M_20_CL_2000001   \n",
       "\n",
       "                                 alias_tissue2 alias_tissue3 alias_tissue4  \\\n",
       "0     ali-mortazavi:016_B6J_10F_03_NTR_0000750           NaN           NaN   \n",
       "1     ali-mortazavi:017_B6J_10M_03_NTR_0000750           NaN           NaN   \n",
       "2     ali-mortazavi:018_B6J_10F_03_NTR_0000750           NaN           NaN   \n",
       "3     ali-mortazavi:019_B6J_10M_03_NTR_0000750           NaN           NaN   \n",
       "4     ali-mortazavi:020_B6J_10F_03_NTR_0000750           NaN           NaN   \n",
       "...                                        ...           ...           ...   \n",
       "1051                                       NaN           NaN           NaN   \n",
       "1052                                       NaN           NaN           NaN   \n",
       "1053                                       NaN           NaN           NaN   \n",
       "1054                                       NaN           NaN           NaN   \n",
       "1055                                       NaN           NaN           NaN   \n",
       "\n",
       "         plate       Protocol Chemistry bc1_well well_type  ... Age_weeks  \\\n",
       "0     igvf_003  Parse_WT_Mega        v2       A1    Single  ...        10   \n",
       "1     igvf_003  Parse_WT_Mega        v2       A2    Single  ...        10   \n",
       "2     igvf_003  Parse_WT_Mega        v2       A3    Single  ...        10   \n",
       "3     igvf_003  Parse_WT_Mega        v2       A4    Single  ...        10   \n",
       "4     igvf_003  Parse_WT_Mega        v2       A5    Single  ...        10   \n",
       "...        ...            ...       ...      ...       ...  ...       ...   \n",
       "1051  igvf_013  Parse_WT_Mega        v2      D12    Single  ...        10   \n",
       "1052  igvf_013  Parse_WT_Mega        v2      E12    Single  ...        10   \n",
       "1053  igvf_013  Parse_WT_Mega        v2      F12    Single  ...        10   \n",
       "1054  igvf_013  Parse_WT_Mega        v2      G12    Single  ...        10   \n",
       "1055  igvf_013  Parse_WT_Mega        v2      H12    Single  ...        10   \n",
       "\n",
       "      Age_days Body_weight_g Estrus_cycle Dissection_date Dissection_time  \\\n",
       "0         72.0          21.1     Diestrus        10/27/22         9:01 AM   \n",
       "1         72.0          26.3          NaN        10/27/22        11:44 AM   \n",
       "2         72.0          21.0    Proestrus        10/27/22         9:38 AM   \n",
       "3         72.0          29.0          NaN        10/27/22        12:12 PM   \n",
       "4         72.0          18.5     Diestrus        10/27/22        10:14 AM   \n",
       "...        ...           ...          ...             ...             ...   \n",
       "1051       NaN          25.9          NaN         7/24/23         1:38 PM   \n",
       "1052       NaN          20.5          NaN         7/24/23        11:10 AM   \n",
       "1053       NaN          26.0          NaN         7/24/23         1:58 PM   \n",
       "1054       NaN          18.9          NaN         7/24/23        11:41 AM   \n",
       "1055       NaN          25.2          NaN         7/24/23         2:21 PM   \n",
       "\n",
       "     Tissue_weight_mg  Multiplexed_Genotype_1 Multiplexed_Genotype_2  \\\n",
       "0               141.0                     NaN                    NaN   \n",
       "1               141.0                     NaN                    NaN   \n",
       "2               143.0                     NaN                    NaN   \n",
       "3               141.0                     NaN                    NaN   \n",
       "4               149.0                     NaN                    NaN   \n",
       "...               ...                     ...                    ...   \n",
       "1051              NaN                     NaN                    NaN   \n",
       "1052              NaN                     NaN                    NaN   \n",
       "1053              NaN                     NaN                    NaN   \n",
       "1054              NaN                     NaN                    NaN   \n",
       "1055              NaN                     NaN                    NaN   \n",
       "\n",
       "     mult_genotype  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              NaN  \n",
       "...            ...  \n",
       "1051           NaN  \n",
       "1052           NaN  \n",
       "1053           NaN  \n",
       "1054           NaN  \n",
       "1055           NaN  \n",
       "\n",
       "[768 rows x 33 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.well_type=='Single']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f6522c0-4022-4f22-b0fa-cb69c9dd73f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79e618da-7688-4c0a-9d0a-a3c1e523b4ef",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14e0a4f3-50b0-41e9-8037-42256565c001",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21773711-efda-4ce3-9fa7-a6ce3d494c00",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "06984b21-472b-45d6-8f23-27656c434320",
   "metadata": {},
   "source": [
    "## 231006 tissue level agg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "25a25620-acf9-474c-9e27-3dd24cc4f87c",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_tsv = '../configs/test_2.tsv'\n",
    "sample_csv = '../configs/sample_metadata.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "30493f32-2dc6-417d-9730-525ffbb5a8d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = parse_config(config_tsv)\n",
    "sample_df = pd.read_csv(sample_csv)\n",
    "wc = {'tissue': 'PBMC'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "426ba7e8-639c-40a4-9959-ae0d808c05f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tissue_adatas(df, sample_df, wc, cfg_entry):\n",
    "    \n",
    "    # limit to input tissue\n",
    "    temp_sample = sample_df.copy(deep=True)\n",
    "    print(len(temp_sample.index))\n",
    "    temp_sample = temp_sample.loc[temp_sample.Tissue==wc['tissue']]\n",
    "    print(len(temp_sample.index))\n",
    "    \n",
    "    # merge this stuff in with the fastq df\n",
    "    fastq_df = df.copy(deep=True)\n",
    "    temp = fastq_df.merge(sample_df, on='plate', how='inner')\n",
    "    \n",
    "    # get the plate / subpool / sample info for this tissue\n",
    "    plates = temp.plate().tolist()\n",
    "    subpools = temp.subpool.tolist()\n",
    "    samples = temp.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "260dcbe9-ed5d-4b90-8f30-0d2d176f0347",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1056\n",
      "96\n",
      "['igvf_013']\n",
      "1\n",
      "['Sublibrary_2', 'Sublibrary_3']\n",
      "2\n",
      "['016_B6J_10F_20', '017_B6J_10M_20', '018_B6J_10F_20', '019_B6J_10M_20', '020_B6J_10F_20', '021_B6J_10M_20', '024_B6J_10F_20', '025_B6J_10M_20', '066_NODJ_10F_20', '067_NODJ_10M_20', '070_NODJ_10F_20', '069_NODJ_10M_20', '072_NODJ_10F_20', '073_NODJ_10M_20', '074_NODJ_10F_20', '075_NODJ_10M_20', '026_AJ_10F_20', '029_AJ_10M_20', '028_AJ_10F_20', '031_AJ_10M_20', '032_AJ_10F_20', '033_AJ_10M_20', '034_AJ_10F_20', '035_AJ_10M_20', '076_PWKJ_10F_20', '077_PWKJ_10M_20', '078_PWKJ_10F_20', '079_PWKJ_10M_20', '080_PWKJ_10F_20', '083_PWKJ_10M_20', '082_PWKJ_10F_20', '085_PWKJ_10M_20', '036_129S1J_10F_20', '037_129S1J_10M_20', '038_129S1J_10F_20', '041_129S1J_10M_20', '040_129S1J_10F_20', '043_129S1J_10M_20', '044_129S1J_10F_20', '045_129S1J_10M_20', '086_CASTJ_10F_20', '087_CASTJ_10M_20', '090_CASTJ_10F_20', '091_CASTJ_10M_20', '092_CASTJ_10F_20', '093_CASTJ_10M_20', '094_CASTJ_10F_20', '095_CASTJ_10M_20', '058_WSBJ_10F_20', '057_WSBJ_10M_20', '060_WSBJ_10F_20', '059_WSBJ_10M_20', '062_WSBJ_10F_20', '061_WSBJ_10M_20', '096_WSBJ_10F_20', '063_WSBJ_10M_20', '046_NZOJ_10F_20', '047_NZOJ_10M_20', '048_NZOJ_10F_20', '049_NZOJ_10M_20', '050_NZOJ_10F_20', '051_NZOJ_10M_20', '052_NZOJ_10F_20', '053_NZOJ_10M_20', '194_B6CASTF1J_10F_20', '193_B6CASTF1J_10M_20', '196_B6CASTF1J_10F_20', '199_B6CASTF1J_10M_20', '202_B6CASTF1J_10F_20', '201_B6CASTF1J_10M_20', '204_B6CASTF1J_10F_20', '203_B6CASTF1J_10M_20', '216_B6PWKF1J_10F_20', '207_B6PWKF1J_10M_20', '218_B6PWKF1J_10F_20', '209_B6PWKF1J_10M_20', '220_B6PWKF1J_10F_20', '211_B6PWKF1J_10M_20', '222_B6PWKF1J_10F_20', '213_B6PWKF1J_10M_20', '224_B6WSBF1J_10F_20', '223_B6WSBF1J_10M_20', '232_B6WSBF1J_10F_20', '225_B6WSBF1J_10M_20', '234_B6WSBF1J_10F_20', '227_B6WSBF1J_10M_20', '236_B6WSBF1J_10F_20', '229_B6WSBF1J_10M_20', '160_B6NODF1J_10F_20', '159_B6NODF1J_10M_20', '164_B6NODF1J_10F_20', '163_B6NODF1J_10M_20', '166_B6NODF1J_10F_20', '165_B6NODF1J_10M_20', '168_B6NODF1J_10F_20', '167_B6NODF1J_10M_20']\n",
      "96\n"
     ]
    }
   ],
   "source": [
    "get_tissue_adatas(df, sample_df, wc, 'test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "441a3bc1-9572-44f4-b23a-d813c9aa06e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6d2e307-2ecd-493c-b26c-4bc1e3a0ade3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f450780-abe4-4b81-81c2-69fd167f176d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e966f39-b100-49b8-b726-7db09dda18d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "047bc5e2-20a5-4c92-a9dc-28066bda74e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ef6f79f-a389-4cdb-bd02-d4a66eabe774",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af1ee36-641f-45eb-a5ae-170c971b2c2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43786610-f8ff-49cb-8695-9942d7d197c6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aa84c2d-42a3-41ed-a18f-1abea81a83d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35877b31-c814-4389-8852-8496cfe70334",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f7f8f3e1-69c1-40f3-81e1-2e7922bb72c8",
   "metadata": {},
   "source": [
    "## 231006 more bc stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d48e5d0-9bd3-4933-a209-494e00af5c8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "kit = 'WT_mega'\n",
    "chemistry = 'v2'\n",
    "bc = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9ee1ef46-aa20-4e5a-8911-af55019d4b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = get_bcs(bc, kit, chemistry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c9cb0cf6-1b7f-49ee-94e9-ec0837fa9a35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bc3</th>\n",
       "      <th>well</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AACGTGAT</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACATCG</td>\n",
       "      <td>A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATGCCTAA</td>\n",
       "      <td>A3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AGTGGTCA</td>\n",
       "      <td>A4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACCACTGT</td>\n",
       "      <td>A5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>GAACAGGC</td>\n",
       "      <td>H8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>GACAGTGC</td>\n",
       "      <td>H9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>GAGTTAGC</td>\n",
       "      <td>H10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>GATGAATC</td>\n",
       "      <td>H11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GCCAAGAC</td>\n",
       "      <td>H12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         bc3 well\n",
       "0   AACGTGAT   A1\n",
       "1   AAACATCG   A2\n",
       "2   ATGCCTAA   A3\n",
       "3   AGTGGTCA   A4\n",
       "4   ACCACTGT   A5\n",
       "..       ...  ...\n",
       "91  GAACAGGC   H8\n",
       "92  GACAGTGC   H9\n",
       "93  GAGTTAGC  H10\n",
       "94  GATGAATC  H11\n",
       "95  GCCAAGAC  H12\n",
       "\n",
       "[96 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db2e91f4-4a23-4bf8-8296-8ff4ea074692",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad26a70d-597e-4580-b681-adfaf6bf1283",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dda81263-0542-45bd-93ac-776e7eb26cd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "kit = 'WT_mega'\n",
    "chemistry = 'v2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d38d1dbb-82e3-4d0e-aa29-f63d47bbcffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc_df = get_bc1_matches(kit, chemistry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "91f41866-2413-436b-9c6c-24a1e7e1b09a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bc1_dt</th>\n",
       "      <th>well</th>\n",
       "      <th>bc1_randhex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CATTCCTA</td>\n",
       "      <td>A1</td>\n",
       "      <td>CATCATCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CTTCATCA</td>\n",
       "      <td>A2</td>\n",
       "      <td>CTGCTTTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCTATATC</td>\n",
       "      <td>A3</td>\n",
       "      <td>CTAAGGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACATTTAC</td>\n",
       "      <td>A4</td>\n",
       "      <td>GCTTATAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTTAGCT</td>\n",
       "      <td>A5</td>\n",
       "      <td>TCTGATCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>ATTAGGCT</td>\n",
       "      <td>H8</td>\n",
       "      <td>GTGTGTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>GCCTTTCA</td>\n",
       "      <td>H9</td>\n",
       "      <td>TATGCTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>ATTCTAGG</td>\n",
       "      <td>H10</td>\n",
       "      <td>ATGGTGTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>CCTTACAT</td>\n",
       "      <td>H11</td>\n",
       "      <td>GAATAATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>ACATTTGG</td>\n",
       "      <td>H12</td>\n",
       "      <td>CCTCTGTG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      bc1_dt well bc1_randhex\n",
       "0   CATTCCTA   A1    CATCATCC\n",
       "1   CTTCATCA   A2    CTGCTTTG\n",
       "2   CCTATATC   A3    CTAAGGGA\n",
       "3   ACATTTAC   A4    GCTTATAG\n",
       "4   ACTTAGCT   A5    TCTGATCC\n",
       "..       ...  ...         ...\n",
       "91  ATTAGGCT   H8    GTGTGTGT\n",
       "92  GCCTTTCA   H9    TATGCTTC\n",
       "93  ATTCTAGG  H10    ATGGTGTT\n",
       "94  CCTTACAT  H11    GAATAATG\n",
       "95  ACATTTGG  H12    CCTCTGTG\n",
       "\n",
       "[96 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db732421-8638-45bb-a800-5768ea843bc4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a613c7f-9821-46c6-b8b4-b0e43210b643",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "fa997d1b-fad9-465c-b80c-03919d6b7aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = '../configs/test_2.tsv'\n",
    "\n",
    "def parse_config(fname):\n",
    "    df = pd.read_csv(fname, sep='\\t')\n",
    "    df['path'] = df.fastq.str.rsplit('/', n=2, expand=True)[0]+'/'\n",
    "    df['path2'] = df.fastq.str.rsplit('/', n=1, expand=True)[0]+'/'\n",
    "    df['r2_fastq'] = df.fastq.str.replace('_R1_', '_R2_')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "011fa960-f85e-4474-a3ad-f32ef6ada46f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_subpool_fastqs(wc, df, how, read=None):\n",
    "    \"\"\"\n",
    "    Get list of fastqs from the same subpool. Can\n",
    "    either return as a Python list of strings or a\n",
    "    formatted string list read to pass to a shell cmd.\n",
    "\n",
    "    Parameters:\n",
    "        how (str): {'str', 'list'}. 'list' will return\n",
    "            Python list of str var. 'str' will return\n",
    "            Python string\n",
    "    \"\"\"\n",
    "    temp = df.copy(deep=True)\n",
    "    temp = temp.loc[(temp.plate==wc['plate'])&\\\n",
    "                    (temp.subpool==wc['subpool'])]\n",
    "\n",
    "    if how == 'list':\n",
    "        reads = [read for i in range(len(temp.index))]\n",
    "        return expand(expand(config['raw']['fastq'],\n",
    "                        zip,\n",
    "                        sample=temp['sample'].tolist(),\n",
    "                        lane=temp['lane'].tolist(),\n",
    "                        allow_missing=True),\n",
    "                        read=read,\n",
    "                        plate=wc['plate'],\n",
    "                        subpool=wc['subpool'])\n",
    "\n",
    "    elif how == 'str':\n",
    "        r1s = expand(expand(config['raw']['fastq'],\n",
    "                        zip,\n",
    "                        sample=temp['sample'].tolist(),\n",
    "                        lane=temp['lane'].tolist(),\n",
    "                        allow_missing=True),\n",
    "                        read='R1',\n",
    "                        plate=wc['plate'],\n",
    "                        subpool=wc['subpool'])\n",
    "        r2s = expand(expand(config['raw']['fastq'],\n",
    "                        zip,\n",
    "                        sample=temp['sample'].tolist(),\n",
    "                        lane=temp['lane'].tolist(),\n",
    "                        allow_missing=True),\n",
    "                        read='R2',\n",
    "                        plate=wc['plate'],\n",
    "                        subpool=wc['subpool'])\n",
    "        fastq_str = ''\n",
    "        for r1, r2 in zip(r1s, r2s):\n",
    "            fastq_str+=f' {r1} {r2}'\n",
    "        return fastq_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "98d24c0b-fc44-4162-a9fb-c0fcef62f1c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_df_info(wc, df, col):\n",
    "    temp = df.copy(deep=True)\n",
    "    temp = temp.loc[(temp.plate==wc['plate'])&\\\n",
    "                    (temp.subpool==wc['subpool'])&\\\n",
    "                    (temp['sample']==wc['sample'])&\\\n",
    "                    (temp.lane==wc['lane'])]\n",
    "    assert len(temp.index) == 1\n",
    "    return temp[col].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "7b95c82c-3f0a-46bf-9f4e-d63bdcd45910",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = parse_config(fname)\n",
    "wc = {'plate': 'igvf_013', 'subpool': 'Sublibrary_2', 'sample': 'S1', 'lane': 'L001', 'read': 'R1'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "886ff73f-858c-4c56-8ddb-d7af26136130",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/dfs7/samlab/seyedam/IGVF/igvf_013/nova1/Sublibrary_2_S1_L001_R2_001.fastq.gz'"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_df_info(wc, df, 'r2_fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "8634714e-15e1-451f-816c-dac8c1837556",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['S1', 'S1', 'S1', 'S1']\n",
      "['L001', 'L002', 'L003', 'L004']\n",
      "R2\n",
      "igvf_013\n",
      "Sublibrary_2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['fastq/igvf_013_Sublibrary_2_S1_L001_R2_001.fastq.gz',\n",
       " 'fastq/igvf_013_Sublibrary_2_S1_L002_R2_001.fastq.gz',\n",
       " 'fastq/igvf_013_Sublibrary_2_S1_L003_R2_001.fastq.gz',\n",
       " 'fastq/igvf_013_Sublibrary_2_S1_L004_R2_001.fastq.gz']"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_subpool_fastqs(wc, df, 'list', read='R2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "44942758-cd04-4d72-9a6a-9ebc1207bfe8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['S1', 'S1', 'S1', 'S1']\n",
      "['L001', 'L002', 'L003', 'L004']\n",
      "None\n",
      "igvf_013\n",
      "Sublibrary_2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "' fastq/igvf_013_Sublibrary_2_S1_L001_R1_001.fastq.gz fastq/igvf_013_Sublibrary_2_S1_L001_R2_001.fastq.gz fastq/igvf_013_Sublibrary_2_S1_L002_R1_001.fastq.gz fastq/igvf_013_Sublibrary_2_S1_L002_R2_001.fastq.gz fastq/igvf_013_Sublibrary_2_S1_L003_R1_001.fastq.gz fastq/igvf_013_Sublibrary_2_S1_L003_R2_001.fastq.gz fastq/igvf_013_Sublibrary_2_S1_L004_R1_001.fastq.gz fastq/igvf_013_Sublibrary_2_S1_L004_R2_001.fastq.gz'"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_subpool_fastqs(wc, df, 'str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "55e480d4-f109-4805-b0b6-7151049360cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/dfs7/samlab/seyedam/IGVF/igvf_005/nova1/Sublibrary_2_S1_L001_R2_001.fastq.gz',\n",
       " '/dfs7/samlab/seyedam/IGVF/igvf_005/nova1/Sublibrary_2_S1_L002_R2_001.fastq.gz']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.r2_fastq.tolist()[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1574aa71-47e8-44bf-be8c-0b9a53d42fbd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
